File: 0042. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        int len = height.size();
        int max = 0, max_i = 0;
        for(int i =0; i<len; i++){
            if(height[i]>=max){
                max = height[i];
                max_i = i;
            }
        }
        if(max == 0) return 0;
        
        
        int l = 0;
        int water = 0;
        
        // Left side of max
        for(int i=0; i<=max_i; i++){
            
            if(height[i]>=height[l]){
                l = i;
            }
            
            else{
                
                water+=height[l]-height[i];
            }
        }
               
        int r = len-1;
        // Right side of max
        for(int i = len-1; i>=max_i; i--){
            
            if(height[i]>=height[r]){
                r = i;
            }
            
            else{
                water+=height[r]-height[i];
            }
        }
        
        return water;
     
    }
};

File: 0042. Trapping Rain Water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        
        n = len(height)
        i = 0
        j = n - 1
        left_max = height[i]
        right_max = height[j]
        water = 0
        while i < j:
            if height[i] < height[j]:
                
                if height[i] >= left_max:
                    left_max = height[i]
                else:
                    water += left_max-height[i]
                i += 1
                
            else:
                
                if height[j] >= right_max:
                    right_max = height[j]
                else:
                    water += right_max - height[j]
                j -= 1
        
        return water 
                    
